### PR TITLE
Builder patter refactoring.

### PR DESCRIPTION
**Pull request by AlexandrouR**
Kotlin code generation feature.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -45,6 +45,6 @@ docs/build
 .DS_Store
 
 .settings
-.project
+.baseProject
```
Comment (cfelde) [inline .gitignore]: This shouldn't have changed

File: README.rst [hunk 2]
```diff
@@ -22,7 +22,7 @@ About
 =====
 The Web3j command line tools enable developers to interact with blockchains more easily. The Web3j command line tools allow allow you to use some of the key functionality of web3j from your terminal, including:
 
-* New project creation
+* New baseProject creation
```
Comment (cfelde) [inline README.rst]: This shouldn't have changed

File: build.gradle [hunk 3]
```diff
@@ -34,7 +34,7 @@ apply {
         download {
             src "https://raw.githubusercontent.com/web3j/build-tools/master/gradle/$buildScript/build.gradle"
             dest "$rootDir/gradle/$buildScript/build.gradle"
-            overwrite true
+            overwrite false
```
Comment (cfelde) [inline build.gradle]: This shouldn't have changed

File: gradle/repositories/build.gradle [hunk 4]
```diff
@@ -1,6 +1,7 @@
 repositories {
     mavenCentral()
     jcenter()
+    mavenLocal()
```
Comment (cfelde) [inline gradle/repositories/build.gradle]: This shouldn't have changed

File: gradlew [hunk 5]
```diff
@@ -180,7 +180,7 @@ APP_ARGS=$(save "$@")
 # Collect all arguments for the java command, following the shell quoting and substitution rules
 eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS"
 
-# by default we should be in the correct project dir, but when run from Finder on Mac, the cwd is wrong
+# by default we should be in the correct baseProject dir, but when run from Finder on Mac, the cwd is wrong
```
Comment (cfelde) [inline gradlew]: This shouldn't have changed

File: src/main/java/org/web3j/console/project/java/UnitTestCreator.java [hunk 6]
```diff
@@ -59,12 +60,12 @@ public static void main(String[] args) {
...
     }
 
-    void generate() throws IOException {
+    public void generate() throws IOException {
         List<Class> compiledClasses = new ClassProvider(new File(wrapperPath)).getClasses();
         compiledClasses.forEach(
                 compiledClass -> {
                     try {
-                        new UnitClassGenerator(
+                        new JavaClassGenerator(
```
Comment (cfelde) [inline src/main/java/org/web3j/console/project/java/UnitTestCreator.java]: Is this change right? Seems odd to go from unit to java?

File: src/main/resources/gradlew.template [hunk 7]
```diff
@@ -162,7 +162,7 @@ APP_ARGS=$(save "$@")
 # Collect all arguments for the java command, following the shell quoting and substitution rules
 eval set -- $DEFAULT_JVM_OPTS $JAVA_OPTS $GRADLE_OPTS "\"-Dorg.gradle.appname=$APP_BASE_NAME\"" -classpath "\"$CLASSPATH\"" org.gradle.wrapper.GradleWrapperMain "$APP_ARGS"
 
-# by default we should be in the correct project dir, but when run from Finder on Mac, the cwd is wrong
+# by default we should be in the correct baseProject dir, but when run from Finder on Mac, the cwd is wrong
```
Comment (cfelde) [inline src/main/resources/gradlew.template]: This shouldn't have changed

File: src/test/resources/keyfiles/readme.txt [hunk 8]
```diff
@@ -1,4 +1,4 @@
 Generated using Geth for the sole purpose of testing key management.
 
-Feel free to do as you please with these files, as they're not used for any of the project
+Feel free to do as you please with these files, as they're not used for any of the baseProject
```
Comment (cfelde) [inline src/test/resources/keyfiles/readme.txt]: This shouldn't have changed

File: .gitignore (hunk 1 above)
Comment (AlexandrouR) [reply .gitignore]: fixed

File: README.rst (hunk 2 above)
Comment (AlexandrouR) [reply README.rst]: fixed

File: gradlew (hunk 5 above)
Comment (AlexandrouR) [reply gradlew]: fixed

File: src/main/resources/gradlew.template (hunk 7 above)
Comment (AlexandrouR) [reply src/main/resources/gradlew.template]: fixed

File: src/test/resources/keyfiles/readme.txt (hunk 8 above)
Comment (AlexandrouR) [reply src/test/resources/keyfiles/readme.txt]: fixed

File: src/main/java/org/web3j/console/project/BaseProject.java [hunk 9]
```diff
@@ -119,37 +117,23 @@ private void createFatJar(String pathToDirectory) throws IOException, Interrupte
         }
     }
 
-    private void generateTopLevelDirectories(ProjectStructure projectStructure) {
+    protected void generateTopLevelDirectories(ProjectStructure projectStructure) {
         projectStructure.createMainDirectory();
+        System.out.println(projectStructure.getMainPath());
```
Comment (cfelde) [inline src/main/java/org/web3j/console/project/BaseProject.java]: Something to clean up

File: src/main/java/org/web3j/console/project/ProjectCreator.java [hunk 10]
```diff
@@ -19,10 +19,13 @@
 import java.util.List;
 import java.util.Optional;
 
+import org.jetbrains.annotations.NotNull;
```
Comment (cfelde) [inline src/main/java/org/web3j/console/project/ProjectCreator.java]: Do we want this dependency?

File: src/main/java/org/web3j/console/project/ProjectImporter.java [hunk 11]
```diff
@@ -15,8 +15,12 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.jetbrains.annotations.NotNull;
```
Comment (cfelde) [inline src/main/java/org/web3j/console/project/ProjectImporter.java]: Do we want this dependency?

File: src/main/java/org/web3j/console/project/kotlin/KotlinProject.java [hunk 12]
```diff
@@ -107,9 +107,9 @@ public KotlinTemplateProvider getTemplateProvider() {
         }
 
         if (withSampleCode) {
-            templateBuilder.withMainJavaClass("Kotlin.template");
+            templateBuilder.withMainKotlinClass("Kotlin.template");
         } else {
-            templateBuilder.withMainJavaClass("EmptyTemplate.java");
+            templateBuilder.withMainKotlinClass("EmptyKotlin.template");
```
Comment (cfelde) [inline src/main/java/org/web3j/console/project/kotlin/KotlinProject.java]: For Java, can we follow a similar naming standard for the template?

Currently it seems to be called EmptyTemplate.java, so rename it to EmptyJava.template I think.

File: src/main/resources/EmptyKotlin.template [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package <package_name>;
+
+class  <project_name> {
```
Comment (cfelde) [inline src/main/resources/EmptyKotlin.template]: Fix formatting, double space.

File: src/main/java/org/web3j/console/project/kotlin/KotlinProject.java (hunk 12 above)
Comment (AlexandrouR) [reply src/main/java/org/web3j/console/project/kotlin/KotlinProject.java]: done

File: src/main/resources/EmptyKotlin.template (hunk 13 above)
Comment (AlexandrouR) [reply src/main/resources/EmptyKotlin.template]: fixed

File: src/main/java/org/web3j/console/project/BaseProject.java (hunk 9 above)
Comment (AlexandrouR) [reply src/main/java/org/web3j/console/project/BaseProject.java]: removed.

File: src/main/java/org/web3j/console/project/java/UnitTestCreator.java (hunk 6 above)
Comment (AlexandrouR) [reply src/main/java/org/web3j/console/project/java/UnitTestCreator.java]: fixed

File: gradle/repositories/build.gradle (hunk 4 above)
Comment (AlexandrouR) [reply gradle/repositories/build.gradle]: This will be removed when web3j is released.

File: build.gradle (hunk 3 above)
Comment (AlexandrouR) [reply build.gradle]: this will be changed back when web3j is released

File: src/main/java/org/web3j/console/project/ProjectCreator.java (hunk 10 above)
Comment (AlexandrouR) [reply src/main/java/org/web3j/console/project/ProjectCreator.java]: Annotation was removed.

File: src/main/resources/Kotlin.template [hunk 14]
```diff
@@ -15,23 +15,25 @@ import java.io.IOException
 import java.nio.file.Files
 import java.nio.file.Paths
 
-class  <project_name> {
+fun main(args: Array<String>) {
+    <project_name>().start(args)
+
```
Comment (cfelde) [inline src/main/resources/Kotlin.template]: Empty line

File: src/main/resources/Kotlin.template [hunk 15]
```diff
@@ -15,23 +15,25 @@ import java.io.IOException
...
-                } catch (e: Exception) {
-                    log.info(e.message)
-                }
-
+             val credentials: Credentials? = loadCredentials("<wallet_name>")
+             val web3j: Web3j? = createWeb3jService("")
+             val helloWorld: HelloWorld? = deployHelloWorld(web3j, credentials, DefaultGasProvider())
+             callGreetMethod(helloWorld)
+         } catch (e: Exception) {
+             log.info(e.message)
```
Comment (cfelde) [inline src/main/resources/Kotlin.template]: This is bad practice, catching an exception and then basically swallowing it. Do we need this catch at all?

File: src/main/resources/build.gradleImport.template [hunk 16]
```diff
@@ -33,7 +34,10 @@ dependencies {
     implementation("org.web3j:web3j-evm:$web3jVersion")
     testImplementation 'org.junit.jupiter:junit-jupiter-api:5.3.1'
     testRuntimeOnly 'org.junit.jupiter:junit-jupiter-engine:5.3.1'
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
+
```
Comment (cfelde) [inline src/main/resources/build.gradleImport.template]: Empty line

Comment (AlexandrouR) [reply src/main/resources/build.gradleImport.template]: Fixed

**Pull request merged**